### PR TITLE
mrc-3001 only return state from run for given index

### PR DESCRIPTION
**Pull request by hillalex**
`mod$state` still returns the full state, and `mod$update_state` still expects the full state, but `mod$run` only returns state for indices set in `mod$set_index` - not sure if it will be useful to be able to update and retrieve state based on the index as well?

### CODE REVIEW

File: inst/include/mode/r/mode.hpp [hunk 1]
```diff
@@ -52,6 +52,55 @@ cpp11::sexp stats_array(const std::vector<size_t>& dat,
   return ret;
 }
 
+int r_index_check(int x, int max) {
```
Comment (hillalex) [inline inst/include/mode/r/mode.hpp]: these helpers lifted straight from dust. wonder if it's time to put utils in a separate file

Comment (richfitz) [reply inst/include/mode/r/mode.hpp]: yeah, I think that would be good

File: tests/testthat/test-interface.R [hunk 2]
```diff
@@ -23,6 +23,20 @@ test_that("Returned state has correct dimensions", {
...
+  gen <- mode(path, quiet = TRUE)
+  pars <- list(r1 = 0.1, r2 = 0.2, K1 = 100, K2 = 100)
+  n_particles <- 10
+  mod <- gen$new(pars, 1, n_particles)
+  mod$set_index(1)
+  res <- mod$solve(2)
+  expect_equal(dim(res), c(1, n_particles))
+
+  state <- mod$state()
+  expect_identical(res, state[1, , drop = F])
```
Comment (richfitz) [inline tests/testthat/test-interface.R]:
```suggestion
  expect_identical(res, state[1, , drop = FALSE])
```

Traumatically, `F <- TRUE` is valid while `FALSE <- TRUE` is a syntax error.

File: tests/testthat/test-interface.R [hunk 3]
```diff
@@ -23,6 +23,20 @@ test_that("Returned state has correct dimensions", {
...
+})
```
Comment (richfitz) [inline tests/testthat/test-interface.R:118]: Can you expand the testing here to check that we can use an index of length > 1 and error if given out of range values (0 or too large)

Comment (richfitz) [reply tests/testthat/test-interface.R:118]: Also, I think that setting an index of `NULL` should wipe the index - that should be tested too.

Edit: looks like dust does not support that actually, but it should!

File: inst/template/mode.R.template [hunk 4]
```diff
@@ -33,6 +34,12 @@
       mode_{{name}}_time(private$ptr_)
     },
 
+    set_index = function(index) {
```
Comment (richfitz) [inline inst/template/mode.R.template:37]: For compatibility with dust, can you add a trivial getter here too?

```
index = function() {
  private$index_
}
```

File: inst/template/mode.R.template [hunk 5]
```diff
@@ -6,7 +6,8 @@
     pars_ = NULL,
     info_ = NULL,
     ptr_ = NULL,
-    n_particles_ = NULL
+    n_particles_ = NULL,
+    index_ = NULL
```
Comment (richfitz) [inline inst/template/mode.R.template:10]: It's not in the diff, but the solve method should be updated to do:

```
    m <- mode_logistic_solve(private$ptr_, end_time)
    rownames(m) <- names(private$index_)
    m
```

which allows a named index to be set (e.g., `set_index(c(I = 2, R = 3))`) and named output to be returned.

We should also rename solve -> run at some point I think...

**Review by richfitz — CHANGES_REQUESTED** [review]

File: inst/template/mode.R.template (hunk 5 above)
Comment (hillalex) [reply inst/template/mode.R.template:10]: https://github.com/mrc-ide/mode/pull/11

**Review by richfitz — APPROVED** [review]

**Pull request merged**